crates/core/tedge_agent/src/tedge_to_te_converter/actor.rs

```diff
@@ -0,0 +1,236 @@
+pub struct ThinEdgeAlarmData {
+    pub text: Option<String>,
+
+    #[serde(skip_serializing_if = "Option::is_none")]
+    #[serde(default)]
+    #[serde(with = "time::serde::rfc3339::option")]
+    pub time: Option<Timestamp>,
+
+    #[serde(default)]
+    pub severity: String,
+
+    #[serde(flatten)]
+    pub alarm_data: HashMap<String, Value>,
+}
+
+pub struct TedgetoTeConverterActor {
+    messages: SimpleMessageBox<MqttMessage, MqttMessage>,
+}
+
+#[async_trait]
+impl Actor for TedgetoTeConverterActor {
+    fn name(&self) -> &str {
+        "TedgetoTeConverterActor"
+    }
+
+    async fn run(&mut self) -> Result<(), RuntimeError> {
+        while let Some(message) = self.messages.recv().await {
+            {
+                self.process_mqtt_message(message).await?;
+            }
+        }
+        Ok(())
+    }
+}
+
+impl TedgetoTeConverterActor {
+    pub fn new(messages: SimpleMessageBox<MqttMessage, MqttMessage>) -> Self {
+        Self { messages }
+    }
+
+    // Todo: Convert command topics
+    //tedge/commands/res/software/list	te/<identifier>/cmd/software_list/<cmd_id>
+    //tedge/commands/res/software/update	te/<identifier>/cmd/software_update/<cmd_id>
+    //tedge/+/commands/res/config_snapshot	te/<identifier>/cmd/config_snapshot/<cmd_id>
+    //tedge/+/commands/res/config_update	te/<identifier>/cmd/config_update/<cmd_id>
+    //tedge/+/commands/res/firmware_update	te/<identifier>/cmd/firmware_update/<cmd_id>
+    //tedge/commands/res/control/restart	te/<identifier>/cmd/restart/<cmd_id>
+    async fn process_mqtt_message(
+        &mut self,
+        message: MqttMessage,
+    ) -> Result<(), TedgetoTeConverterError> {
+        match message.topic.clone() {
+            topic if topic.name.starts_with("tedge/measurements") => {
+                self.convert_measurement(message).await?;
+            }
+            topic if topic.name.starts_with("tedge/events") => {
+                self.convert_event(message).await?;
+            }
+            topic if topic.name.starts_with("tedge/alarms") => {
+                self.convert_alarm(message).await?;
+            }
+            topic if topic.name.starts_with("tedge/health") => {
+                self.convert_health_status_message(message).await?;
+            }
+            _ => {}
+        }
+        Ok(())
+    }
+
+    // tedge/measurements -> te/device/main///m/
+    // tedge/measurements/child -> te/device/child///m/
+    async fn convert_measurement(
+        &mut self,
+        message: MqttMessage,
+    ) -> Result<(), TedgetoTeConverterError> {
+        let cid = get_child_id_from_measurement_topic(&message.topic.name)?;
+        let te_topic = match cid {
+            Some(cid) => Topic::new_unchecked(format!("te/device/{cid}///m/").as_str()),
+            None => Topic::new_unchecked("te/device/main///m/"),
+        };
+
+        let msg = MqttMessage::new(&te_topic, message.payload_str()?).with_qos(message.qos);
```

Comment:
As the source message is owned, it's simpler to edit it in place.

```suggestion
       message.topic = te_topic;
```